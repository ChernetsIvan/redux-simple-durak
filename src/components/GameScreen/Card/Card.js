import React from "react";
import PropTypes from "prop-types";

import "./Card.css";

import clubSuit from "./../../../assets/card-suits/club.png";
import diamondSuit from "./../../../assets/card-suits/diamond.png";
import heartSuit from "./../../../assets/card-suits/heart.png";
import spadeSuit from "./../../../assets/card-suits/spade.png";

import { uniqueId } from "lodash-es";

class Card extends React.Component {
  constructor(props) {
    super(props);

    this.handleClickOnCard = this.handleClickOnCard.bind(this);
  }

  handleClickOnCard(e) {
    this.props.onCardClick(this.props.id);
  }

  render() {
    let suit = this.props.suit.suit.toString();
    let rank = this.props.rank.text.toString();
    let suitImg, sizeText;
    switch (suit) {
      case "Ч": {
        suitImg = heartSuit;
        sizeText = <b className="redText">{rank}</b>;
        break;
      }
      case "К": {
        suitImg = clubSuit;
        sizeText = <b className="blackText">{rank}</b>;
        break;
      }
      case "П": {
        suitImg = spadeSuit;
        sizeText = <b className="blackText">{rank}</b>;
        break;
      }
      case "Б": {
        suitImg = diamondSuit;
        sizeText = <b className="redText">{rank}</b>;
        break;
      }
      default:
        break;
    }

    let classAttr = this.props.bootStrapColClass + " card mb-1";
    let randKey = uniqueId();
    let output = (
      <div key={randKey} onClick={this.handleClickOnCard} className={classAttr}>
        <div className="row">
          <b className="ml-1">{sizeText}</b>
        </div>
        <div className="row">
          <img className="littleSuitImg ml-1" src={suitImg} alt="" />
        </div>
        <div className="row justify-content-center">
          <img className="bigSuitImg my-2" src={suitImg} alt="" />
        </div>
        <div className="row justify-content-end">
          <img className="littleSuitImg rotated mr-1" src={suitImg} alt="" />
        </div>
        <div className="row justify-content-end">
          <b className="rotated mr-1">{sizeText}</b>
        </div>
      </div>
    );
    return output;
  }
}

export default Card;

Card.propTypes = {
  id: PropTypes.number,
  key: PropTypes.number.isRequired,
  rank: PropTypes.shape({
    text: PropTypes.text,
    cardValue: PropTypes.number
  }).isRequired,
  suit: PropTypes.shape({
    suit: PropTypes.string
  }).isRequired,
  onCardClick: PropTypes.func,
  bootStrapColClass: PropTypes.string.isRequired
};
