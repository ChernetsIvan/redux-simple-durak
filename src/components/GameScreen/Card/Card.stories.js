import React from "react";

import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";

import "./../../../../public/libs/bootstrap/css/bootstrap.min.css";
//а js-файлы (и bootstrap'a) подключены в .storybook/preview-head.html

import Card from "./Card";

storiesOf("Card", module)
  .addDecorator(story => (
    <div className="container">
      <div className="row">{story()}</div>
    </div>
  ))
  .add("with hidden=FALSE", () => (
    <Card
      id={1}
      rank={new Object({ text: "6", cardValue: 1 })}
      suit={new Object({ suit: "К" })}
      onCardClick={action("Card was clicked!")}
      bootStrapColClass={"col-1"}
      hidden={false}
    />
  ))
  .add("yet another", () => (
    <Card
      id={2}
      rank={new Object({ text: "Д", cardValue: 7 })}
      suit={new Object({ suit: "Б" })}
      onCardClick={action("Card was clicked!")}
      bootStrapColClass={"col-1"}
      hidden={false}
    />
  ))
  .add("with hidden=TRUE", () => (
    <Card id={3} bootStrapColClass="col-1" hidden={true} />
  ));
