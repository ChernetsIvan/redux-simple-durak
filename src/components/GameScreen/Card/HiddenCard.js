import React from "react";
import PropTypes from "prop-types";

import "./Card.css";

//Отрисовывает карту рубашкой кверху
export const HiddenCard = ({ bootStrapColClass }) => {
  let classAttr = bootStrapColClass + " card mb-1 shirt";
  return <div className={classAttr} />;
};

HiddenCard.propTypes = {
  bootStrapColClass: PropTypes.string.isRequired
};
