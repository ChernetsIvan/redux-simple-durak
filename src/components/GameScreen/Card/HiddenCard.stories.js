import React from "react";

import { storiesOf } from "@storybook/react";

import HiddenCard from "./HiddenCard";

storiesOf("HiddenCard", module)
  .addDecorator(story => (
    <div className="container">
      <div className="row">{story()}</div>
    </div>
  ))
  .add("example", () => <HiddenCard bootStrapColClass={"col-1"} />);
