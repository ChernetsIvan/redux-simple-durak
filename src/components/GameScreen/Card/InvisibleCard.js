import React from "react";

import { HiddenCard } from "./HiddenCard";

import { uniqueId } from "lodash-es";

//Применяется AI.js, Player.js и Field.js - когда во входном массиве нет карт:
//чтобы "не скакала" разметка.
export const InvisibleCard = () => {
  let randKey = uniqueId();
  return (
    <div className="container invisible">
      <div className="row">
        <HiddenCard key={randKey} bootStrapColClass="col-1" />
      </div>
    </div>
  );
};
