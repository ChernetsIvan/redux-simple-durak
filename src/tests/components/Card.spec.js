import React from "react";
import renderer from "react-test-renderer";
import Enzyme, { shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-15";
import Card from "./../../components/GameScreen/Card/Card";

describe("Card snapshot testing", () => {
  test("Card with suit К", () => {
    const emptyFunc = () => {};
    const component = renderer.create(
      <Card
        id={1}
        rank={{ text: "6", cardValue: 1 }}
        suit={{ suit: "К" }}
        onCardClick={emptyFunc}
        bootStrapColClass={"col-1"}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("Card with suit Ч", () => {
    const emptyFunc = () => {};
    const component = renderer.create(
      <Card
        id={2}
        rank={{ text: "6", cardValue: 1 }}
        suit={{ suit: "Ч" }}
        onCardClick={emptyFunc}
        bootStrapColClass={"col-1"}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("Card with suit Б", () => {
    const emptyFunc = () => {};
    const component = renderer.create(
      <Card
        id={2}
        rank={{ text: "6", cardValue: 1 }}
        suit={{ suit: "Б" }}
        onCardClick={emptyFunc}
        bootStrapColClass={"col-1"}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("Card with suit П", () => {
    const emptyFunc = () => {};
    const component = renderer.create(
      <Card
        id={2}
        rank={{ text: "6", cardValue: 1 }}
        suit={{ suit: "П" }}
        onCardClick={emptyFunc}
        bootStrapColClass={"col-1"}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  test("Card with not existing suit", () => {
    const emptyFunc = () => {};
    const component = renderer.create(
      <Card
        id={2}
        rank={{ text: "6", cardValue: 1 }}
        suit={{ suit: "Й" }}
        onCardClick={emptyFunc}
        bootStrapColClass={"col-1"}
      />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

Enzyme.configure({ adapter: new Adapter() });

describe("Test click on Card", () => {
  test("Test click event", () => {
    const mockCallBack = jest.fn();
    const card = shallow(
      <Card
        id={2}
        rank={{ text: "6", cardValue: 1 }}
        suit={{ suit: "П" }}
        onCardClick={mockCallBack}
        bootStrapColClass={"col-1"}
      />
    );
    card.find('[className="col-1 card mb-1"]').simulate("click");
    expect(mockCallBack.mock.calls.length).toEqual(1);
  });
});
