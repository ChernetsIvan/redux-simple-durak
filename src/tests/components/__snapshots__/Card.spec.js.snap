// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card snapshot testing Card with not existing suit 1`] = `
<div
  className="col-1 card mb-1"
  onClick={[Function]}
>
  <div
    className="row"
  >
    <b
      className="ml-1"
    />
  </div>
  <div
    className="row"
  >
    <img
      alt=""
      className="littleSuitImg ml-1"
      src={undefined}
    />
  </div>
  <div
    className="row justify-content-center"
  >
    <img
      alt=""
      className="bigSuitImg my-2"
      src={undefined}
    />
  </div>
  <div
    className="row justify-content-end"
  >
    <img
      alt=""
      className="littleSuitImg rotated mr-1"
      src={undefined}
    />
  </div>
  <div
    className="row justify-content-end"
  >
    <b
      className="rotated mr-1"
    />
  </div>
</div>
`;

exports[`Card snapshot testing Card with suit Б 1`] = `
<div
  className="col-1 card mb-1"
  onClick={[Function]}
>
  <div
    className="row"
  >
    <b
      className="ml-1"
    >
      <b
        className="redText"
      >
        6
      </b>
    </b>
  </div>
  <div
    className="row"
  >
    <img
      alt=""
      className="littleSuitImg ml-1"
      src="diamond.png"
    />
  </div>
  <div
    className="row justify-content-center"
  >
    <img
      alt=""
      className="bigSuitImg my-2"
      src="diamond.png"
    />
  </div>
  <div
    className="row justify-content-end"
  >
    <img
      alt=""
      className="littleSuitImg rotated mr-1"
      src="diamond.png"
    />
  </div>
  <div
    className="row justify-content-end"
  >
    <b
      className="rotated mr-1"
    >
      <b
        className="redText"
      >
        6
      </b>
    </b>
  </div>
</div>
`;

exports[`Card snapshot testing Card with suit К 1`] = `
<div
  className="col-1 card mb-1"
  onClick={[Function]}
>
  <div
    className="row"
  >
    <b
      className="ml-1"
    >
      <b
        className="blackText"
      >
        6
      </b>
    </b>
  </div>
  <div
    className="row"
  >
    <img
      alt=""
      className="littleSuitImg ml-1"
      src="club.png"
    />
  </div>
  <div
    className="row justify-content-center"
  >
    <img
      alt=""
      className="bigSuitImg my-2"
      src="club.png"
    />
  </div>
  <div
    className="row justify-content-end"
  >
    <img
      alt=""
      className="littleSuitImg rotated mr-1"
      src="club.png"
    />
  </div>
  <div
    className="row justify-content-end"
  >
    <b
      className="rotated mr-1"
    >
      <b
        className="blackText"
      >
        6
      </b>
    </b>
  </div>
</div>
`;

exports[`Card snapshot testing Card with suit П 1`] = `
<div
  className="col-1 card mb-1"
  onClick={[Function]}
>
  <div
    className="row"
  >
    <b
      className="ml-1"
    >
      <b
        className="blackText"
      >
        6
      </b>
    </b>
  </div>
  <div
    className="row"
  >
    <img
      alt=""
      className="littleSuitImg ml-1"
      src="spade.png"
    />
  </div>
  <div
    className="row justify-content-center"
  >
    <img
      alt=""
      className="bigSuitImg my-2"
      src="spade.png"
    />
  </div>
  <div
    className="row justify-content-end"
  >
    <img
      alt=""
      className="littleSuitImg rotated mr-1"
      src="spade.png"
    />
  </div>
  <div
    className="row justify-content-end"
  >
    <b
      className="rotated mr-1"
    >
      <b
        className="blackText"
      >
        6
      </b>
    </b>
  </div>
</div>
`;

exports[`Card snapshot testing Card with suit Ч 1`] = `
<div
  className="col-1 card mb-1"
  onClick={[Function]}
>
  <div
    className="row"
  >
    <b
      className="ml-1"
    >
      <b
        className="redText"
      >
        6
      </b>
    </b>
  </div>
  <div
    className="row"
  >
    <img
      alt=""
      className="littleSuitImg ml-1"
      src="heart.png"
    />
  </div>
  <div
    className="row justify-content-center"
  >
    <img
      alt=""
      className="bigSuitImg my-2"
      src="heart.png"
    />
  </div>
  <div
    className="row justify-content-end"
  >
    <img
      alt=""
      className="littleSuitImg rotated mr-1"
      src="heart.png"
    />
  </div>
  <div
    className="row justify-content-end"
  >
    <b
      className="rotated mr-1"
    >
      <b
        className="redText"
      >
        6
      </b>
    </b>
  </div>
</div>
`;
